packages/client/src/authentication/adapters/Credentials.ts

```diff
@@ -0,0 +1,71 @@
+import { DateUtils, invariant, InvariantError } from '@lens-protocol/shared-kernel';
+import jwtDecode, { JwtPayload } from 'jwt-decode';
+
+function isObject(value: unknown): value is object {
+  const type = typeof value;
+  return value != null && (type === 'object' || type === 'function');
+}
+
+type ParsedJwt = {
+  id: string;
+};
+
+function assertIdInJwt(decodedJwt: unknown): asserts decodedJwt is ParsedJwt {
+  if (isObject(decodedJwt) && !('id' in decodedJwt)) {
+    throw new InvariantError('Invalid JWT format.');
+  }
+}
+
+// Threshold in seconds that will mark token as expired even it's still valid
+// Adds some time for all communications that's required to refresh tokens
+const TOKEN_EXP_THRESHOLD = DateUtils.secondsToMs(3);
+
+export class Credentials {
+  readonly address: string;
+
+  constructor(readonly accessToken: string | undefined, readonly refreshToken: string) {
+    const decodedRefreshToken = jwtDecode<JwtPayload>(refreshToken);
+
+    assertIdInJwt(decodedRefreshToken);
+
+    this.address = decodedRefreshToken.id;
+  }
+
+  canRefresh(): boolean {
+    const now = Date.now();
+    const tokenExpDate = this.getTokenExpDate(this.refreshToken);
+
+    return now < tokenExpDate - TOKEN_EXP_THRESHOLD;
+  }
+
+  isExpired(): boolean {
+    const accessToken = this.accessToken;
+
+    if (!accessToken) {
+      return true;
+    }
+
+    const now = Date.now();
+    const tokenExpDate = this.getTokenExpDate(accessToken);
+
+    return now >= tokenExpDate - TOKEN_EXP_THRESHOLD;
+  }
+
+  getAccessTokenExpDate(): number | null {
```

Comment:
It says date but seems a timestamp.

Reply:
that method was removed but I renamed the others to `getTokenExpTimestamp`